crates/sp-domains/src/verification.rs

```diff
@@ -141,29 +147,42 @@ where
     }
 
     let storage_key = SK::block_randomness_key();
-    let storage_proof = randomness_proof.clone();
     let block_randomness = StorageProofVerifier::<Hashing>::verify_and_get_value::<Randomness>(
         &consensus_state_root,
-        storage_proof,
+        randomness_proof.clone(),
         storage_key,
     )
     .map_err(|_| VerificationError::InvalidProof)?;
 
+    let timestamp = StorageProofVerifier::<Hashing>::verify_and_get_value::<Moment>(
+        &consensus_state_root,
+        timestamp_proof.clone(),
+        SK::timestamp_storage_key(),
+    )
+    .map_err(|_| VerificationError::InvalidProof)?;
+
     let shuffling_seed =
         H256::decode(&mut extrinsics_shuffling_seed::<Hashing>(block_randomness).as_ref())
             .map_err(|_| VerificationError::FailedToDecode)?;
-    let ordered_extrinsics = deduplicate_and_shuffle_extrinsics(
+
+    let mut ordered_extrinsics = deduplicate_and_shuffle_extrinsics(
         bundle_extrinsics_digests,
         Randomness::from(shuffling_seed.to_fixed_bytes()),
     );
+
+    let timestamp_extrinsic =
+        ExtrinsicDigest::new::<LayoutV1<BlakeTwo256>>(DE::derive_timestamp_extrinsic(timestamp));
+    ordered_extrinsics.insert(0, timestamp_extrinsic);
```

Comment:
It would probably be better if `deduplicate_and_shuffle_extrinsics` returned `VecDeque` in this case, otherwise this will potentially involve quite some memory copying.